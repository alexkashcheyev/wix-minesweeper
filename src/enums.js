export const actionType = {
    TOGGLE_MENU: 'TOGGLE_MENU',
    TOGGLE_SUPERMAN: 'TOGGLE_SUPERMAN',
    SET_NEW_GAME_PARAMETER: 'SET_NEW_GAME_PARAMETER',
    START_GAME: 'START_GAME',
    TOGGLE_FLAG: 'TOGGLE_FLAG',
    OPEN_CELL: 'OPEN_CELL',
    MOVE_VIEWPORT: 'MOVE_VIEWPORT',
    SET_MESSAGE: 'SET_MESSAGE',
    SET_FIELD: 'SET_FIELD',
    UPDATE_FLAG_COUNT: 'UPDATE_FLAG_COUNT',
    SET_VIEWPORT_OFFSET: 'SET_VIEWPORT_OFFSET',
    SET_GAME_STAGE: 'SET_GAME_STAGE',
    SET_VIEWPORT_SIZE: 'SET_VIEWPORT_SIZE'
}

export const gameStage = {
    NOT_STARTED: 'NOT_STARTED',
    STARTED: 'STARTED',
    WON: 'WON',
    LOST: 'LOST'
}