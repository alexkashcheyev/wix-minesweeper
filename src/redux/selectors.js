export const selectCurrentGameInfo = (state) => state.currentGame.gameInfo;

export const selectCurrentGame = (state) => state.currentGame;

export const selectField = (state) => state.currentGame.field;