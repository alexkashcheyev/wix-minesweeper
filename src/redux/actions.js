import { actionType } from '../enums';

// request to show or hide the menu
export const toggleMenu = (show) => ({
    type: actionType.TOGGLE_MENU,
    payload: { show }
});

// set superman mode to given value
export const toggleSuperman = (superman) => ({
    type: actionType.TOGGLE_SUPERMAN,
    payload: { superman }
});

// set new game parameter, which is 'width', 'height' or 'mines'
// to given value, if valid
export const setNewGameParameter = (key, value) => ({
    type: actionType.SET_NEW_GAME_PARAMETER,
    payload: { key, value }
});

// request starting game
export const startGame = (gameInfo) => ({
    type: actionType.START_GAME,
    payload: { gameInfo }
})

// set or unset flag on given cell
export const toggleFlag = (x, y) => ({
    type: actionType.TOGGLE_FLAG,
    payload: { x, y }
});

// request opening a cell
export const openCell = (x, y) => ({
    type: actionType.OPEN_CELL,
    payload: { x, y }
})

// request to move the viewport to see different part of the field
export const moveViewport = (dx, dy) => ({
    type: actionType.MOVE_VIEWPORT,
    payload: { dx, dy }
})

// show/hide the alert with given data
export const setMessage = (visible, severity='info', title='', content='') => ({
    type: actionType.SET_MESSAGE,
    payload: { severity, visible, title, content }
})

// set the game field as a reflection
// of user's actions
export const setField = (field) => ({
    type: actionType.SET_FIELD,
    payload: { field }
})

// increase flags count by given number
export const changeSetFlags = (delta) => ({
    type: actionType.UPDATE_FLAG_COUNT,
    payload: { delta }
})

// actually set the viewport offset
export const setViewportOffset = (offset) => ({
    type: actionType.SET_VIEWPORT_OFFSET,
    payload: { offset }
})

// move game to another stage
export const setGameStage = (stage) => ({
    type: actionType.SET_GAME_STAGE,
    payload: { stage }
})