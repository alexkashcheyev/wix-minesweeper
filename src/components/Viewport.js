import React, { useEffect, useRef } from 'react';
import { makeStyles, Button, Typography, Paper } from '@material-ui/core';
import GameField from './GameField';
import Minimap from './Minimap'
import { connect } from 'react-redux';
import { gameStage, actionType } from '../enums';
import * as actions from '../redux/actions';
import { validViewportChange } from '../shared/viewport';
import config from '../appconfig';

import KeyboardArrowUpRoundedIcon from '@material-ui/icons/KeyboardArrowUpRounded';
import KeyboardArrowLeftRoundedIcon from '@material-ui/icons/KeyboardArrowLeftRounded';
import KeyboardArrowDownRoundedIcon from '@material-ui/icons/KeyboardArrowDownRounded';
import KeyboardArrowRightRoundedIcon from '@material-ui/icons/KeyboardArrowRightRounded';
import FlagRoundedIcon from '@material-ui/icons/FlagRounded';

const useStyles = makeStyles((theme) => ({
    root: {
        display: 'flex',
        flexDirection: 'row'
    },
    panel : {
        margin: theme.spacing(1),
        flex: '0',
        alignItems: 'center',
        minWidth: config.panelWidth,
        display: 'flex',
        flexDirection: 'column',
        padding: theme.spacing(1)
    },
    stats: {
        marginBottom: theme.spacing(2)
    },
    minimap: {
        height: config.minimapSize,
        width: config.minimapSize,
        display: 'flex',
        justifyContent: 'center',
        alignItems: 'center',
        flexDirection: 'column'
    },
    divider: {
        flex: '1'
    },
    buttons: {
        display: 'grid',
        gridColumnGap: theme.spacing(1),
        gridRowGap: theme.spacing(1),
        gridTemplateAreas: ' ". U ." "L D R" '
    },
    btnUp: {
        gridArea: 'U'
    },
    btnDown: {
        gridArea: 'D'
    },
    btnLeft: {
        gridArea: 'L'
    },
    btnRight: {
        gridArea: 'R'
    }
}));

function Viewport({ dispatch, viewport, gameInfo, stage, flagsSet }) {
    console.log(viewport, gameInfo);
    
    const classes = useStyles();
    
    const changeOffset = (dx, dy) => {
        dispatch(actions.changeOffset(dx, dy));
    }

    const handleKeyPress = (e) => {
        
        let change = false;

        if ( ['KeyW', 'ArrowUp'].indexOf(e.code) >= 0 ) {
            change = { dx: 0 , dy: -1 }
        } else if (['KeyS', 'ArrowDown'].indexOf(e.code) >= 0) {
            change = { dx: 0, dy: 1 }
        } else if (['KeyA', 'ArrowLeft'].indexOf(e.code) >= 0) {
            change = { dx: -1, dy: 0 }
        } else if (['KeyD', 'ArrowRight'].indexOf(e.code) >= 0) {
            change = { dx: 1, dy: 0 }
        }

        if (change) {
            dispatch(actions.changeOffset(change.dx, change.dy))
        }
    }

    useEffect(() => {
        document.addEventListener('keyup', handleKeyPress);
        return () => {
            document.removeEventListener('keyup', handleKeyPress);
        }
    }, [document]);

    let stats = (
        <div className={classes.stats}>
            <Typography color='secondary' variant='h4' component='div'>
                <FlagRoundedIcon /> &times; { gameInfo.mines - flagsSet }
            </Typography>
        </div>
    );

    let panel = (
        <Paper className={classes.panel} variant='outlined'>
            {stats}
        </Paper>
    );
    let key = 0;

    if (viewport.height < gameInfo.height || viewport.width < gameInfo.width) {

        panel =
            (
                <Paper 
                    variant='outlined'
                    className={classes.panel}>

                    {stats}
                    <div className={classes.minimap}>
                        <Minimap 
                            gameInfo={gameInfo} 
                            viewport={viewport} />
                        <Typography variant='caption'>Minimap</Typography>
                    </div>
                    <div className={classes.divider}>
                        &nbsp;
                    </div>
                    <div className={classes.buttons}>
                        <Button
                            key={key++}
                            variant='contained'
                            className={classes.btnUp}
                            disabled={!validViewportChange(viewport, gameInfo, 0, -1)}
                            startIcon={<KeyboardArrowUpRoundedIcon />}
                            onClick={() => changeOffset(0, -1)}>

                            W
                        </Button>
                        <Button
                            key={key++}
                            variant='contained'
                            className={classes.btnDown}
                            disabled={!validViewportChange(viewport, gameInfo, 0, 1)}
                            startIcon={<KeyboardArrowDownRoundedIcon />}
                            onClick={() => changeOffset(0, 1)}>

                            S
                        </Button>
                        <Button
                            key={key++}
                            variant='contained'
                            className={classes.btnLeft}
                            disabled={!validViewportChange(viewport, gameInfo, -1, 0)}
                            startIcon={<KeyboardArrowLeftRoundedIcon />}
                            onClick={() => changeOffset(-1, 0)}>

                            A
                        </Button>
                        <Button
                            key={key++}
                            variant='contained'
                            className={classes.btnRight}
                            disabled={!validViewportChange(viewport, gameInfo, 1, 0)}
                            startIcon={<KeyboardArrowRightRoundedIcon />}
                            onClick={() => changeOffset(1, 0)}>
                            D
                        </Button>
                    </div>
                </Paper>
            )
    }

    const fieldBorder = {
        top     : !validViewportChange(viewport, gameInfo,  0, -1),
        bottom  : !validViewportChange(viewport, gameInfo,  0,  1),
        left    : !validViewportChange(viewport, gameInfo, -1,  0),
        right   : !validViewportChange(viewport, gameInfo,  1,  0)
    }

    return stage === gameStage.NOT_STARTED ? (
        <Typography variant='h4'>
            To start the game, click on the hamburger menu in the left upper corner, enter the desired parameters and click "Start" button.
        </Typography>
    ) : (
        <div className={classes.root}>
            {panel}
            <GameField className={classes.field} border={fieldBorder} />
        </div>
    )
}

const mapStateToProps = (state) => state.currentGame;

export default connect(mapStateToProps)(Viewport);