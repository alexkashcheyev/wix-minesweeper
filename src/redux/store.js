import { createStore, applyMiddleware } from 'redux';
import { gameStage, actionType } from '../enums';
import createSagaMiddleware from 'redux-saga';
import toggleFlagSaga from '../sagas/toggleFlagSaga';
import generateFieldSaga from '../sagas/generateFieldSaga';
import changeOffsetSaga from '../sagas/changeOffsetSaga';
import openCellSaga from '../sagas/openCellSaga';
import { fork, all } from 'redux-saga/effects';
import config from '../appconfig';

const intialState = {
    ui: {
        showMenu: false,
        message: {
            severity: 'info',
            visible: false,
            title: '',
            content: ''
        }
    },
    currentGame: {
        gameInfo: {
            width: 15,
            height: 15,
            mines: 10
        },
        viewport: {
            width: config.viewportWidth,
            height: config.viewportHeight,
            offset: {
                x: 0,
                y: 0
            },
        },
        stage: gameStage.NOT_STARTED,
        superman: true,
        flagsSet: 0,
        field: []
    },
    newGame: {
        gameInfo: {
            width: 300,
            height: 300,
            mines: 1
        }
    }
}

const rootReducer = (state = intialState, action) => {
    switch (action.type) {

        case actionType.TOGGLE_MENU:
            return {
                ...state,
                ui: {
                    ...state.ui,
                    showMenu: action.payload.show
                }
            }

        case actionType.TOGGLE_SUPERMAN:
            return {
                ...state,
                currentGame: {
                    ...state.currentGame,
                    superman: action.payload.superman
                }
            }

        case actionType.SET_NEW_GAME_PARAMETER:
            return {
                ...state,
                newGame: {
                    ...state.newGame,
                    gameInfo: {
                        ...state.newGame.gameInfo,
                        [action.payload.key]: action.payload.value
                    }
                }
            }

        case actionType.START_GAME:
            return {
                ...state,
                currentGame: {
                    ...state.currentGame,
                    ...action.payload,
                    flagsSet: 0,
                    stage: gameStage.LOADING,
                    viewport: {
                        ...state.currentGame.viewport,
                        offset: {
                            x: 0,
                            y: 0
                        }
                    }
                },
            }

        case actionType.SET_MESSAGE: {

            return {
                ...state,
                ui: {
                    ...state.ui,
                    message: action.payload
                }
            }
            
        }

        case actionType.SET_FIELD: {

            return {
                ...state,
                currentGame: {
                    ...state.currentGame,
                    field: action.payload.field
                }
            }
        }

        case actionType.UPDATE_FLAG_COUNT: {
            
            return {
                ...state,
                currentGame: {
                    ...state.currentGame,
                    flagsSet: state.currentGame.flagsSet + action.payload.delta
                }
            }
        }

        case actionType.SET_VIEWPORT_OFFSET: {

            return {
                ...state,
                currentGame: {
                    ...state.currentGame,
                    viewport: {
                        ...state.currentGame.viewport,
                        offset: action.payload.offset                        
                    }
                }
            }
        }

        case actionType.SET_GAME_STAGE: {
            return {
                ...state,
                currentGame: {
                    ...state.currentGame,
                    stage: action.payload.stage
                }
            }
        }

        default: return state;
    }
}

const sagaMiddleware = createSagaMiddleware();

export const store = createStore(
    rootReducer, 
    applyMiddleware(sagaMiddleware)
);

function* rootSaga () {
    yield all([
        fork(toggleFlagSaga),
        fork(generateFieldSaga),
        fork(changeOffsetSaga),
        fork(openCellSaga)
    ]);
}

sagaMiddleware.run(rootSaga);