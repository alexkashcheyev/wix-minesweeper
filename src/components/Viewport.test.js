import React from 'react';
import TestRenderer, { act } from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';

import { initialState } from '../redux/store';
import Viewport from "./Viewport"
import { gameStage } from '../enums';


describe('Viewport component', () => {
    const mockStore = configureStore([]);
    let store;
    let tr;
    let component;
    
    const wrappedJsx = (store) => (
        <Provider store={store}>
            <Viewport />
        </Provider>
    )
    
    const update = () => {
        act(
            () => {
                tr.update(wrappedJsx(store));
            }
        )
        component = tr.toJSON();
    }

    beforeEach(() => {
        store = mockStore(initialState);

        act(() => {
            tr = TestRenderer.create(wrappedJsx(store));
        });

        component = tr.toJSON();
    })    
    
    it('should create', () => {
        expect(component).toBeTruthy();
    });

    it('should show text if game was not started', () => {
        expect(component).toMatchSnapshot();
    });

    it('should show only flag count if game has started, and game field fits', () => {
        store = mockStore({
            ...initialState,
            currentGame: {
                ...initialState.currentGame,
                stage: gameStage.STARTED,
                gameInfo: {
                    ...initialState.currentGame.gameInfo,
                    width: 5,
                    height: 5
                },
            },
            viewport: {
                ...initialState.currentGame.viewport,
                width: 10,
                height: 10
            }
        });

        update();

        expect(component).toMatchSnapshot();
    });

    it('should also show minimap and navigation keys if game does not fit', () => {
        store = mockStore({
            ...initialState,
            currentGame: {
                ...initialState.currentGame,
                stage: gameStage.STARTED,
                gameInfo: {
                    ...initialState.currentGame.gameInfo,
                    width: 10,
                    height: 10
                },
                viewport: {
                    ...initialState.viewport,
                    width: 5,
                    height: 5
                }
            },
        });

        update();

        expect(component).toMatchSnapshot();
    });

})

// for debugging this file:
// node --inspect-brk ./node_modules/jest/bin/jest.js --runInBand viewport.test.js