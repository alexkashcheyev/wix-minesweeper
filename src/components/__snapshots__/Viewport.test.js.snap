// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Viewport component should also show minimap and navigation keys if game does not fit 1`] = `
<div
  className="makeStyles-root-1"
>
  <div
    className="MuiPaper-root makeStyles-panel-2 MuiPaper-outlined MuiPaper-rounded"
  >
    <div
      className="makeStyles-stats-3"
    >
      <div
        className="MuiTypography-root MuiTypography-h4 MuiTypography-colorSecondary"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M14.4 6l-.24-1.2c-.09-.46-.5-.8-.98-.8H6c-.55 0-1 .45-1 1v15c0 .55.45 1 1 1s1-.45 1-1v-6h5.6l.24 1.2c.09.47.5.8.98.8H19c.55 0 1-.45 1-1V7c0-.55-.45-1-1-1h-4.6z"
          />
        </svg>
         × 
        10
      </div>
    </div>
    <div
      className="makeStyles-minimap-4"
    >
      <div
        className="MuiPaper-root makeStyles-root-15 makeStyles-root-17 MuiPaper-outlined"
      >
        <div
          className="makeStyles-viewport-16 makeStyles-viewport-18"
        >
           
        </div>
      </div>
      <span
        className="MuiTypography-root MuiTypography-caption"
      >
        Minimap
      </span>
    </div>
    <div
      className="makeStyles-divider-5"
    >
       
    </div>
    <div
      className="makeStyles-buttons-6"
    >
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-btnUp-8 Mui-disabled Mui-disabled"
        disabled={true}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={-1}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          <span
            className="MuiButton-startIcon MuiButton-iconSizeMedium"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M8.12 14.71L12 10.83l3.88 3.88c.39.39 1.02.39 1.41 0 .39-.39.39-1.02 0-1.41L12.7 8.71a.9959.9959 0 00-1.41 0L6.7 13.3c-.39.39-.39 1.02 0 1.41.39.38 1.03.39 1.42 0z"
              />
            </svg>
          </span>
          W
        </span>
      </button>
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-btnDown-9"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          <span
            className="MuiButton-startIcon MuiButton-iconSizeMedium"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M8.12 9.29L12 13.17l3.88-3.88c.39-.39 1.02-.39 1.41 0 .39.39.39 1.02 0 1.41l-4.59 4.59c-.39.39-1.02.39-1.41 0L6.7 10.7a.9959.9959 0 010-1.41c.39-.38 1.03-.39 1.42 0z"
              />
            </svg>
          </span>
          S
        </span>
        <span
          className="MuiTouchRipple-root"
        />
      </button>
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-btnLeft-10 Mui-disabled Mui-disabled"
        disabled={true}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={-1}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          <span
            className="MuiButton-startIcon MuiButton-iconSizeMedium"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M14.71 15.88L10.83 12l3.88-3.88c.39-.39.39-1.02 0-1.41a.9959.9959 0 00-1.41 0L8.71 11.3c-.39.39-.39 1.02 0 1.41l4.59 4.59c.39.39 1.02.39 1.41 0 .38-.39.39-1.03 0-1.42z"
              />
            </svg>
          </span>
          A
        </span>
      </button>
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-btnRight-11"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          <span
            className="MuiButton-startIcon MuiButton-iconSizeMedium"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M9.29 15.88L13.17 12 9.29 8.12a.9959.9959 0 010-1.41c.39-.39 1.02-.39 1.41 0l4.59 4.59c.39.39.39 1.02 0 1.41L10.7 17.3c-.39.39-1.02.39-1.41 0-.38-.39-.39-1.03 0-1.42z"
              />
            </svg>
          </span>
          D
        </span>
        <span
          className="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
  <div
    className="makeStyles-fieldWrapper-7"
  >
    <div
      className="makeStyles-root-12 makeStyles-root-19"
    />
  </div>
</div>
`;

exports[`Viewport component should show only flag count if game has started, and game field fits 1`] = `
<div
  className="makeStyles-root-1"
>
  <div
    className="MuiPaper-root makeStyles-panel-2 MuiPaper-outlined MuiPaper-rounded"
  >
    <div
      className="makeStyles-stats-3"
    >
      <div
        className="MuiTypography-root MuiTypography-h4 MuiTypography-colorSecondary"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M14.4 6l-.24-1.2c-.09-.46-.5-.8-.98-.8H6c-.55 0-1 .45-1 1v15c0 .55.45 1 1 1s1-.45 1-1v-6h5.6l.24 1.2c.09.47.5.8.98.8H19c.55 0 1-.45 1-1V7c0-.55-.45-1-1-1h-4.6z"
          />
        </svg>
         × 
        10
      </div>
    </div>
  </div>
  <div
    className="makeStyles-fieldWrapper-7"
  >
    <div
      className="makeStyles-root-12 makeStyles-root-14"
    />
  </div>
</div>
`;

exports[`Viewport component should show text if game was not started 1`] = `
<h4
  className="MuiTypography-root MuiTypography-h4"
>
  To start the game, click on the hamburger menu in the left upper corner, enter the desired parameters and click "Start" button.
</h4>
`;
